qubes/storage/zfs.py

```diff
@@ -0,0 +1,2281 @@
+        self,
+        volume: Volume,
+        log: logging.Logger,
+    ) -> int:
+        """Get the current size of the volume."""
+        assert dataset_in_root(volume, self.root)
+        async with self.cache.lock:
+            cached = self.cache.get(volume, "volsize")
+            if cached is not None:
+                return cast(int, cached)
+            ret = int(await self._get_prop_async(volume, "volsize", log))
+            self.cache.set(volume, "volsize", ret)
+            return ret
+
+    async def set_volume_dirty_async(
+        self,
+        volume: Volume,
+        dirty: bool,
+        log: logging.Logger,
+    ) -> None:
+        """
+        Mark the volume as dirty.
+
+        Usually used when a VM backed by this volume is started.
+        """
+        assert dataset_in_root(volume, self.root)
+        async with self.cache.lock:
+            self.cache.invalidate_recursively(volume, "org.qubes:dirty")
+            val = "on" if dirty else "off"
+            await self._set_prop_async(volume, "org.qubes:dirty", val, log)
+            self.cache.set(volume, "org.qubes:dirty", dirty)
+
+    def is_volume_dirty(
+        self,
+        volume: Volume,
+        log: logging.Logger,
+    ) -> bool:
+        """Sync version of `is_volume_dirty()`.  Results not cached."""
+        assert dataset_in_root(volume, self.root)
+        v = self._get_prop(volume, "org.qubes:dirty", log)
+        ret = v == "on"
+        return ret
+
+    async def is_volume_dirty_async(
+        self,
+        volume: Volume,
+        log: logging.Logger,
+    ) -> bool:
+        """
+        Returns true if the volume is dirty.
+
+        The volume must exist.
+        """
+        assert dataset_in_root(volume, self.root)
+        async with self.cache.lock:
+            dirty = self.cache.get(volume, "org.qubes:dirty")
+            if dirty is not None:
+                return cast(bool, dirty)
+            v = await self._get_prop_async(volume, "org.qubes:dirty", log)
+            ret = v == "on"
+            self.cache.set(volume, "org.qubes:dirty", ret)
+            return ret
+
+    def get_volume_snapshots(
+        self,
+        volume: Volume,
+        log: logging.Logger,
+    ) -> List[VolumeSnapshot]:
+        """
+        Get all snapshots of the volume.
+
+        The volume must exist.
+
+        The return value is a list of `VolumeSnapshot`.
+
+        The result of this method is never cached.
+        """
+        assert dataset_in_root(volume, self.root)
+        ret = [
+            VolumeSnapshot.make(s.split("@")[0], s.split("@")[1])
+            for s in zfs(
+                "list", "-Hp", "-o", "name", "-t", "snapshot", volume, log=log
+            ).splitlines()
+            if s.strip()
+        ]
+        return ret
+
+    async def get_volume_snapshots_async(
+        self,
+        volume: Volume,
+        log: logging.Logger,
+        lock: bool = True,
+    ) -> List[VolumeSnapshot]:
```

Comment:
There are two calls to this method (in the most up-to-date pushed code):

* L#1212 calls this within an already-locked context.  Thus lock=False.
* L#1860 calls this within an unlocked context.  Thus lock=True.

Thus, this is necessary to prevent a recursive deadlock.

Reply:
Nonetheless, will be fixed in the next push.